File: nutrition/src/main/java/com/example/nutrition/service/UserService.java
package com.example.nutrition.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.nutrition.dao.UserDao;
import com.example.nutrition.entity.Code;
import com.example.nutrition.entity.User;
import com.example.nutrition.service.UserService;

@Service
public class UserService {
    @Autowired
    private UserDao userDao;

    public int signUp(User user) {
        if (userDao.isRegistered(user.getUsername()) != null)
            return Code.SAME_USERNAME;
        else if (userDao.insertUser(user.getUsername(), user.getPassword()) > 0)
            return Code.SAVE_OK;
        else
            return Code.SAVE_ERROR;
    }

    public int signIn(User user) {
        if (userDao.isRegistered(user.getUsername()) == null)
            return Code.UNFOUND_USER;
        if (user.getPassword().equals(userDao.getPassword(user.getUsername())) == false)
            return Code.WRONG_PASSWORD;
        else
            return Code.GET_OK;
    }

    public int updateUsername(String oldUsername, String newUsername) {
        if (userDao.isRegistered(oldUsername) == null)
            return Code.UNFOUND_USER;
        else if (userDao.isRegistered(newUsername) != null)
            return Code.SAME_USERNAME;
        else if (userDao.updateUsername(oldUsername, newUsername) > 0)
            return Code.UPDATE_OK;
        else
            return Code.UPDATE_ERROR;
    }

    public int updatePassword(String username, String oldPassword, String newPassword) {
        if (userDao.isRegistered(username) == null)
            return Code.UNFOUND_USER;
        else if (oldPassword.equals(userDao.getPassword(username)) == false)
            return Code.WRONG_PASSWORD;
        else if (userDao.updatePassword(username, newPassword) > 0)
            return Code.UPDATE_OK;
        else
            return Code.UPDATE_ERROR;
    }

    public int updateDetails(User user) {
        if (userDao.isRegistered(user.getUsername()) == null)
            return Code.UNFOUND_USER;
        else if (userDao.updateDetails(user) > 0)
            return Code.UPDATE_OK;
        else
            return Code.UPDATE_ERROR;
    }

    public User getUser(String username) {
        return userDao.getByUsername(username);
    }
}

File: spitts/src/service/UserService.go
package service

import (
	"spitts/src/dao"
	"spitts/src/entity"
	"spitts/src/global"
)

func SignUp(user entity.User) entity.Result {
	if dao.IsRegistered(user.Username) {
		return entity.Result{Code: global.USED_USERNAME}
	}

	err := dao.InsertUser(user.Username, user.Password, 0)
	if err != nil {
		return entity.Result{Code: global.UPDATE_ERROR}
	}

	return entity.Result{Code: global.SUCCESS}
}

func SignIn(user entity.User) entity.Result {
	if !dao.IsRegistered(user.Username) {
		return entity.Result{Code: global.USER_NOTFOUND}
	}

	if !dao.CheckPassword(user.Username, user.Password) {
		return entity.Result{Code: global.WRONG_PASSWORD}
	}

	return_user := dao.SelectUser(user.Username)
	return_user.Password = ""
	if return_user.Username == "" {
		return entity.Result{Code: global.QUERY_ERROR}
	}

	return entity.Result{Code: global.SUCCESS, Object: return_user}
}

func UpdateUsername(old_username string, new_username string) entity.Result {
	if !dao.IsRegistered(old_username) {
		return entity.Result{Code: global.USER_NOTFOUND}
	} else if dao.IsRegistered(new_username) {
		return entity.Result{Code: global.USED_USERNAME}
	}

	err := dao.UpdateUsername(old_username, new_username)
	if err != nil {
		return entity.Result{Code: global.UPDATE_ERROR}
	}

	return entity.Result{Code: global.SUCCESS}
}

func UpdatePassword(username string, old_password string, new_password string) entity.Result {
	if !dao.CheckPassword(username, old_password) {
		return entity.Result{Code: global.WRONG_PASSWORD}
	}

	err := dao.UpdatePassword(username, new_password)
	if err != nil {
		return entity.Result{Code: global.UPDATE_ERROR}
	}

	return entity.Result{Code: global.SUCCESS}
}
